frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/common/SlideOut.java

```diff
@@ -31,125 +32,100 @@
+public class SlideOut {
 
   private final Project project;
   private final String branchToSlideOutName;
   @Nullable
   private final String currentBranchNameIfManaged;
   private final BranchLayout branchLayout;
   private final GitRepository gitRepository;
-  private final BaseEnhancedGraphTable graphTable;
-  private final @UI Runnable doInUIThreadWhenReady;
 
   public static final String DELETE_LOCAL_BRANCH_ON_SLIDE_OUT_GIT_CONFIG_KEY = "machete.slideOut.deleteLocalBranch";
 
-  public SlideOutBackgroundable(String title, IManagedBranchSnapshot branchToSlideOut,
+  public SlideOut(IManagedBranchSnapshot branchToSlideOut,
       GitRepository gitRepository,
       @Nullable IManagedBranchSnapshot currentBranchNameIfManaged,
-      BranchLayout branchLayout,
-      BaseEnhancedGraphTable graphTable,
-      @UI Runnable doInUIThreadWhenReady) {
-    super(gitRepository.getProject(), title);
+      BranchLayout branchLayout) {
     this.project = gitRepository.getProject();
     this.branchToSlideOutName = branchToSlideOut.getName();
     this.currentBranchNameIfManaged = currentBranchNameIfManaged != null ? currentBranchNameIfManaged.getName() : null;
     this.branchLayout = branchLayout;
     this.gitRepository = gitRepository;
-    this.graphTable = graphTable;
-    this.doInUIThreadWhenReady = doInUIThreadWhenReady;
 
     LOG.debug(() -> "Entering: branchToSlideOut = ${branchToSlideOutName}");
     LOG.debug("Refreshing repository state");
   }
 
-  @Override
   @ContinuesInBackground
-  @UIThreadUnsafe
-  public void run(ProgressIndicator indicator) {
+  public void run(@UI Runnable doInUIThreadWhenReady) {
     val slideOutBranchIsCurrent = branchToSlideOutName.equals(currentBranchNameIfManaged);
     if (slideOutBranchIsCurrent) {
       LOG.debug("Skipping (optional) local branch deletion because it is equal to current branch");
       slideOutBranch(branchToSlideOutName);
-      graphTable.queueRepositoryUpdateAndModelRefresh(doInUIThreadWhenReady);
       VcsNotifier.getInstance(project).notifySuccess(/* displayId */ null,
           /* title */ "",
-          fmt(getString("action.GitMachete.BaseSlideOutAction.notification.title.slide-out-success.of-current.HTML"),
+          getString("action.GitMachete.BaseSlideOutAction.notification.title.slide-out-success.of-current.HTML").fmt(
               branchToSlideOutName));
 
     } else {
       val root = gitRepository.getRoot();
-      val shouldDelete = getDeleteLocalBranchOnSlideOutGitConfigValue(root);
-      if (shouldDelete == null) {
-        ModalityUiUtil.invokeLaterIfNeeded(ModalityState.NON_MODAL,
-            () -> suggestBranchDeletion(branchToSlideOutName, doInUIThreadWhenReady));
-      } else {
-        handleBranchDeletionDecision(branchToSlideOutName, shouldDelete);
-        ModalityUiUtil.invokeLaterIfNeeded(ModalityState.NON_MODAL, doInUIThreadWhenReady);
-      }
+      getDeleteLocalBranchOnSlideOutGitConfigValueAndExecute(root, shouldDelete -> {
```

Comment:
That's basically a [_callback hell_](http://callbackhell.com/).

Until IntelliJ (and Java in general) migrates to Futures, it won't get any easier ☹️ `map`/`flatMap` are waaaay easier to manipulate than this 💩 I didn't face even the tenth of race conditions, and didn't need to invent stuff like what's in `BackgroundTaskEnqueuingTestSuite`.

To absolve IntelliJ a bit, the fact is that the strict separation between UI and non-UI thread **might** make using Futures more cumbersome... I haven't ever seen such combination in practice. Mostly used Futures for server-side code, where still some amount of [bulkheading](https://stackoverflow.com/questions/47250947/what-is-bulkheading-in-computer-science) is present... but that doesn't come even close to the strictness of UI threading.